src/tasks/task-angular-momentum-equality.cpp

```diff
@@ -0,0 +1,174 @@
+    }
+
+    int TaskAMEquality::dim() const
+    {
+      return (int)m_mask.sum();
+    }
+    const Vector3 & TaskAMEquality::mask() const
+    {
+      return m_mask;
+    }
+
+    void TaskAMEquality::mask(const Vector3 & m)
+    {
+      //TODO: ggory15
+      m_mask = m;
+      const Vector::Index dim = static_cast<Vector::Index>(m.sum());
+      Matrix S = Matrix::Zero(dim, m_robot.nv());
+      m_activeAxes.resize(dim);
+      unsigned int j=0;
+      for(unsigned int i=0; i<m.size(); i++)
+        if(m(i)!=0.0)
+        {
+          assert(m(i)==1.0);
+          m_activeAxes(j) = i;
+          j++;
+        }
+      m_constraint.resize((unsigned int)dim, m_robot.nv());
+    }
+
+
+    const Vector3 & TaskAMEquality::Kp(){ return m_Kp; }
+
+    const Vector3 & TaskAMEquality::Kd(){ return m_Kd; }
+
+    void TaskAMEquality::Kp(ConstRefVector Kp)
+    {
+      assert(Kp.size()==3);
+      m_Kp = Kp;
+    }
+
+    void TaskAMEquality::Kd(ConstRefVector Kd)
+    {
+      assert(Kd.size()==3);
+      m_Kd = Kd;
+    }
+
+    void TaskAMEquality::setReference(const TrajectorySample & ref)
+    {
+      m_ref = ref;
+    }
+
+    const TrajectorySample & TaskAMEquality::getReference() const
+    {
+      return m_ref;
+    }
+
+    const Vector3 & TaskAMEquality::getDesireddMomentum() const
+    {
+      return m_dL_des;
+    }
+
+    Vector3 TaskAMEquality::getdMomentum(ConstRefVector dv) const
+    {
+      return m_constraint.matrix()*dv - m_drift;
+    }
+
+    const Vector3 & TaskAMEquality::momentum_error() const
+    {
+      return m_L_error;
+    }
+
+    const Vector3 & TaskAMEquality::momentum() const
+    {
+      return m_L;
+    }
+    const Vector & TaskAMEquality::momentum_ref() const
+    {
+      return m_ref.pos;
+    }
+
+    const Vector & TaskAMEquality::dmomentum_ref() const
+    {
+      return m_ref.acc;
+    }
+
+    const ConstraintBase & TaskAMEquality::getConstraint() const
+    {
+      return m_constraint;
+    }
+
+    const ConstraintBase & TaskAMEquality::compute(const double ,
+                                                    ConstRefVector q,
+                                                    ConstRefVector v,
+                                                    const Data & data)
+    {
+      // Compute errors
+      // Get momentum jacobian
+      Data data_ccrba = data;
+      pinocchio::ccrba(m_robot.model(), data_ccrba, q, v);
+      const Matrix6x & J_am = data_ccrba.Ag;
+      m_L = J_am.bottomRows(3) * v;
+      m_L_error = m_L - m_ref.pos;
+
+      m_dL_des = - m_Kp.cwiseProduct(m_L_error)
+                + m_ref.acc;
+
+#ifndef NDEBUG
+      std::cout<<m_name<<" errors: "<<m_L_error.norm()<<" "
+        <<m_dL_error.norm()<<std::endl;
+#endif
+      // Del Prete's quick and dirty way to compute drift
+      // compute momentum Jacobian at next time step assuming zero acc
+      double dt = 1e-3;
+      const Vector & q_next = pinocchio::integrate(m_robot.model(), q, dt*v);
```

Comment:
This trick was acceptable in my personal python library, but it's too dirty for TSID. I think you should compute the drift in a proper way. Maybe @jcarpent can suggest you which function of Pinocchio to use to do that.